### PR TITLE
[427] - [FE] Integrate Fetch, and Upload Files Functionality

### PR DESCRIPTION
**Pull request by ww-jermaine**
## Issue Link
- https://app.asana.com/0/1203011167276287/1203247946500427/f

## Definition of Done
- [x] A member of a project can see project-related files when clicking the `Files` tab.
- [x] A product owner of a project can upload project-related files when clicking the `Upload a file` button.
- [x] A team leader of a project can upload project-related files when clicking the `Upload a file` button. 
- [x] Added permissions.
- [x] Added Skeleton Table Loading.  
- [x] Fixed issue with loading. 

## Pre-condition
cli: `cd api`
- php artisan migrate:fresh --seed

cli: `cd client`
- yarn dev || npm run dev
- go to `project/:id/files` to see project-related files and upload a file.

## Expected Output
- The page should display all existing project-related files.
- The user should be able to upload a file and the page should display the recently uploaded file by the user.

## Screenshots/Recordings

As a product owner

https://user-images.githubusercontent.com/109291819/199669705-313e2ce0-d49d-4cf5-8622-fe1adfdb1f07.mp4

As a team lead

https://user-images.githubusercontent.com/109291819/199669723-a9b80a7e-05b6-4fe9-baac-0b8532f7e0af.mp4

As a member

https://user-images.githubusercontent.com/109291819/199680282-cc917d7e-23ed-40d7-9988-3e2364caa867.mp4

Loading

[Slackana _ Overview (1).webm](https://user-images.githubusercontent.com/109291819/199690855-d3f72ead-7376-4edf-8337-63f99638c04f.webm)

### CODE REVIEW

File: client/src/components/molecules/FileListSkeleton/index.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import React, { FC } from 'react'
+import LineSkeleton from '~/components/atoms/Skeletons/LineSkeleton'
+
+type Props = {
+  length: number
+}
+
+const FileListSkeleton: FC<Props> = (props): JSX.Element => {
+  const { length } = props
+
```
Comment (abduljalilpalala) [inline client/src/components/molecules/FileListSkeleton/index.tsx]:
```suggestion
const FileListSkeleton: FC<Props> = ({ length }): JSX.Element => {


```

File: client/src/components/molecules/PaginationSkeleton/index.tsx [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import React, { FC } from 'react'
+import LineSkeleton from '~/components/atoms/Skeletons/LineSkeleton'
+
```
Comment (abduljalilpalala) [inline client/src/components/molecules/PaginationSkeleton/index.tsx]:
```suggestion
import React, { FC } from 'react'

import LineSkeleton from '~/components/atoms/Skeletons/LineSkeleton'

```

File: client/src/pages/team/[id]/files.tsx [hunk 3]
```diff
@@ -1,14 +1,29 @@
...
 import { Filename } from '~/shared/types'
-import { files } from '~/shared/jsons/filesData'
 import FileList from '~/components/molecules/FileList'
 import Pagination from '~/components/atoms/Pagination'
+import PaginationSkeleton from '~/components/molecules/PaginationSkeleton'
 import ProjectLayout from '~/components/templates/ProjectLayout'
+import { darkToaster } from '~/utils/darkToaster'
+
+import { useFileMethods } from '~/hooks/fileMethods'
+import { useAppSelector } from '~/hooks/reduxSelector'
```
Comment (abduljalilpalala) [inline client/src/pages/team/[id]/files.tsx]:
```suggestion
import { darkToaster } from '~/utils/darkToaster'
import { useFileMethods } from '~/hooks/fileMethods'
import { useAppSelector } from '~/hooks/reduxSelector'
```

File: client/src/pages/team/[id]/files.tsx [hunk 4]
```diff
@@ -32,15 +47,22 @@ const Files: NextPage = (): JSX.Element => {
...
+    })
     setIsOpen(!isOpen)
   }
 
   /*
    * Handle Update Filename
    */
   const handleUpdateFilename = async (data: Filename): Promise<void> => {
-    alert(JSON.stringify(data, null, 2))
+    // alert(JSON.stringify(data, null, 2))
```
Comment (abduljalilpalala) [inline client/src/pages/team/[id]/files.tsx]: Remove comment.

File: client/src/pages/team/[id]/files.tsx [hunk 5]
```diff
@@ -100,7 +135,35 @@ const Files: NextPage = (): JSX.Element => {
  * - List of Files
  * - Upload Files button
  */
-const FileHeader = (): JSX.Element => {
+const FileHeader = (props: any): JSX.Element => {
+  const error = useAppSelector((state) => state.file.error)
```
Comment (abduljalilpalala) [inline client/src/pages/team/[id]/files.tsx]: Change `any` to it's appropriate type

File: client/src/redux/files/fileSlice.ts [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+      if (response === 'File uploaded successfully') {
+        return {
+          message: response,
+          files: formattedFiles
+        }
+      }
+      return thunkAPI.rejectWithValue({ status: 422, content: response, files: formattedFiles })
+    } catch (error) {
+      return thunkAPI.rejectWithValue(catchError(error))
+    }
```
Comment (abduljalilpalala) [inline client/src/redux/files/fileSlice.ts]: Why invoking rejectWithValue in the `try` section? `catchError` method in the catch section already handles 422 error. Try checking the implementation if it suits your desired output.

File: client/src/redux/files/fileSlice.ts [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        payload.projectId,
+        payload.fileId,
+        payload.name
+      )
+      if (response === 'File updated successfully') {
+        const files = await fileService.getProjectFiles(payload.projectId)
+        const formattedFiles = formatFiles(files)
+        return formattedFiles
+      }
+    } catch (error) {
```
Comment (abduljalilpalala) [inline client/src/redux/files/fileSlice.ts:82]: Do you mean `response.data.message === 'File updated successfully'`?

File: client/src/redux/files/fileSlice.ts [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  'file/deleteProjectFile',
+  async (payload: { projectId: number; fileId: string }, thunkAPI) => {
+    try {
+      const response = await fileService.deleteProjectFile(payload.projectId, payload.fileId)
+      if (response === 'File deleted successfully') {
+        const files = await fileService.getProjectFiles(payload.projectId)
+        const formattedFiles = formatFiles(files)
+        return formattedFiles
+      }
+    } catch (error) {
```
Comment (abduljalilpalala) [inline client/src/redux/files/fileSlice.ts:98]: Same with my previous comment.

**Review by abduljalilpalala — COMMENTED** [review]
Please check comments.

File: client/src/redux/files/fileSlice.ts (hunk 7 above)
Comment (ww-jermaine) [reply client/src/redux/files/fileSlice.ts:82]: From our previous review of implemented axios methods yesterday, `renameProjectFile` returns a `Promise<string>`

```
const renameProjectFile = async (
  projectId: number,
  fileId: string,
  name: string
): Promise<string>
```

**Review by abduljalilpalala — APPROVED** [review]
### Approved! ✔️

**Review by rogeliojohnoliverio — APPROVED** [review]
Approved 🙇

**Pull request merged**